d2compiler/compile.go

```diff
@@ -34,534 +32,409 @@ func Compile(path string, r io.RuneReader, opts *CompileOptions) (*d2graph.Graph
 		UTF16: opts.UTF16,
 	})
 	if err != nil {
-		if !errors.As(err, &pe) {
-			return nil, err
-		}
+		return nil, err
 	}
 
-	return compileAST(path, pe, ast)
-}
-
-func compileAST(path string, pe d2parser.ParseError, ast *d2ast.Map) (*d2graph.Graph, error) {
-	g := d2graph.NewGraph(ast)
+	ir, err := d2ir.Compile(ast)
+	if err != nil {
+		return nil, err
+	}
 
-	c := &compiler{
-		path: path,
-		err:  pe,
+	g, err := compileIR(pe, ir)
+	if err != nil {
+		return nil, err
 	}
+	g.AST = ast
+	return g, err
+}
 
-	c.compileKeys(g.Root, ast)
-	if len(c.err.Errors) == 0 {
-		c.validateKeys(g.Root, ast)
+func compileIR(pe d2parser.ParseError, m *d2ir.Map) (*d2graph.Graph, error) {
+	c := &compiler{
+		err: pe,
 	}
-	c.compileEdges(g.Root, ast)
-	// TODO: simplify removeContainer by running before compileEdges
-	c.compileShapes(g.Root)
-	c.validateNear(g)
 
+	g := c.compileLayer(m)
 	if len(c.err.Errors) > 0 {
 		return nil, c.err
 	}
 	return g, nil
 }
 
-type compiler struct {
-	path string
-	err  d2parser.ParseError
-}
+func (c *compiler) compileLayer(ir *d2ir.Map) *d2graph.Graph {
+	g := d2graph.NewGraph()
 
-func (c *compiler) errorf(start d2ast.Position, end d2ast.Position, f string, v ...interface{}) {
-	r := d2ast.Range{
-		Path:  c.path,
-		Start: start,
-		End:   end,
-	}
-	f = "%v: " + f
-	v = append([]interface{}{r}, v...)
-	c.err.Errors = append(c.err.Errors, d2ast.Error{
-		Range:   r,
-		Message: fmt.Sprintf(f, v...),
-	})
+	m := ir.CopyRoot()
+	c.compileMap(g.Root, m)
+	if len(c.err.Errors) == 0 {
+		c.validateKeys(g.Root, m)
+	}
+	c.validateNear(g)
+
+	c.compileLayersField(g, ir, "layers")
```

Comment:
did u mean to use the original `ir` instead of `m`?

Reply:
Yes, `m` doesn't have layers/scenarios/steps to avoid them being created as objects. Scenarios/steps/layers are a purely IR stage thing. I suppose I could add some code to just ignore scenarios/steps/layers keys instead of doing a CopyRoot though.

Reply:
Done, no copy anymore just ignore elsewhere.